blog/_src/posts/2016-06-27-tutorial-using-racket-s-ffi.scrbl

```diff
@@ -0,0 +1,335 @@
+a bitmap object that we'll draw on. The @racket[get-handle] method just extracts a
+low-level Cairo surface value that we can use.
+
+Our first real step is to import the FFI itself:
+
+@examples[#:eval ev #:label #f (require ffi/unsafe)]
+
+As the module name suggests, the FFI is @emph{unsafe} and can cause your Racket process
+to segfault. If you're following along in DrRacket, you will want to save your file
+frequently.
+
+Next, we can load the Cairo library to obtain a @ffitech{foreign-library value}, which
+is a handle that we use to access C values and functions:
+
+@examples[#:eval ev #:label #f
+(define cairo-lib (ffi-lib #f))
+]
+
+Since Cairo has already been loaded by the Racket process because of the
+@racketmodname[racket/gui] import earlier, we can supply @racket[#f] here as an
+argument to @racket[ffi-lib]. Normally you supply the name of a shared
+library file such as @racket["libcairo"]:
+
+@racketblock[
+(define cairo-lib (ffi-lib "libcairo" '(#f 2)))
+]
+
+The last list argument specifies the accepted versions (@racket[#f] allows
+a version-less library). For this post, those details aren't important but
+see the docs on @racket[ffi-lib] if you're curious.
+
+@section{Extracting functions}
+
+Since the Racket FFI is a dynamic interface, we can pull out C functions
+at run-time using the @racket[get-ffi-obj] function. The @racket[get-ffi-obj]
+function takes three arguments:
+
+@itemize[
+  @item{The name of the value as a string (or symbol or bytestring)}
+  @item{a foreign library value, and}
+  @item{a @ffitech{C type}, which is a type description that tells
+        the FFI how to marshall between Racket and C.}
+]
+
+C types are a crucial concept for the FFI. They range from relatively
+simple types like @racket[_int] and @racket[_pointer] to more complicated
+type constructors such as @racket[_enum] and @racket[_fun]. As you probably noticed,
+C types are prefixed with an underscore by convention. You can also define
+your own types by calling @racket[make-ctype] with two functions that
+handle marshalling between C and Racket code.
+
+To make progress with our Cairo code, we need to create a drawing context from
+the surface object @racket[bt-surface] that we defined a while ago. The
+relevant function in the Cairo docs is
+@hyperlink["https://www.cairographics.org/manual/cairo-cairo-t.html#cairo-create"]{@tt{cairo_create}},
+which has the following type signature:
+
+@codeblock[#:keep-lang-line? #f]|{
+#lang honu
+cairo_t * cairo_create (cairo_surface_t *target);
+}|
+
+To use this function from Racket, we will need to create a C type that describes
+its behavior. As you can see, the function takes a pointer to a @tt{cairo_surface_t} and
+returns a pointer to a @tt{cairo_t}. Let's start with a very simple C type
+that matches up with this behavior: @racketblock[(_fun _pointer -> _pointer)]
+Note that the FFI library uses infix arrow notation for its @racket[_fun] type. This
```

Comment:
The order of these sentences ("Note that ..." and "This type ...") confused me. Can we swap them?

(At first I thought the `_fun` type provided very little safety because of its infix arrow notation.)